service/history/mutableStateTaskRefresher.go

```diff
@@ -0,0 +1,394 @@
+		decision.ScheduleID,
+	)
+}
+
+func (r *mutableStateTaskRefresherImpl) refreshTasksForActivity(
+	nowTimestamp int64,
+) error {
+
+	executionInfo := r.mutableState.GetExecutionInfo()
+	pendingActivityInfos := r.mutableState.GetPendingActivityInfos()
+
+Loop:
+	for _, activityInfo := range pendingActivityInfos {
+		// clear all activity timer task mask for later activity timer task re-generation
+		activityInfo.TimerTaskStatus = TimerTaskStatusNone
+
+		if activityInfo.StartedID != common.EmptyEventID {
+			continue Loop
+		}
+
+		scheduleEvent, err := r.eventsCache.getEvent(
+			executionInfo.DomainID,
+			executionInfo.WorkflowID,
+			executionInfo.RunID,
+			activityInfo.ScheduledEventBatchID,
+			activityInfo.ScheduleID,
+			r.mutableState.GetEventStoreVersion(),
+			r.mutableState.GetCurrentBranch(),
+		)
+		if err != nil {
+			return err
+		}
+
+		if err := r.taskGenerator.generateActivityTransferTasks(
+			nowTimestamp,
+			scheduleEvent,
+		); err != nil {
+			return err
+		}
+	}
+
+	tBuilder := newTimerBuilder(r.logger, r.getTimeSource(nowTimestamp))
+	if timerTask := tBuilder.GetActivityTimerTaskIfNeeded(
+		r.mutableState,
+	); timerTask != nil {
+		// no need to set the version, since activity timer task
+		// is just a trigger to check all activities
+		r.mutableState.AddTimerTasks(timerTask)
+	}
+
+	return nil
+}
+
+func (r *mutableStateTaskRefresherImpl) refreshTasksForTimer(
+	nowTimestamp int64,
+) error {
+
+	pendingTimerInfos := r.mutableState.GetPendingTimerInfos()
+
+	for _, timerInfo := range pendingTimerInfos {
+		// clear all timer task mask for later timer task re-generation
+		timerInfo.TaskID = TimerTaskStatusNone
+	}
+
+	tBuilder := newTimerBuilder(r.logger, r.getTimeSource(nowTimestamp))
+	if timerTask := tBuilder.GetUserTimerTaskIfNeeded(
+		r.mutableState,
+	); timerTask != nil {
+		// no need to set the version, since activity timer task
+		// is just a trigger to check all activities
+		r.mutableState.AddTimerTasks(timerTask)
+	}
+
+	return nil
+}
+
+func (r *mutableStateTaskRefresherImpl) refreshTasksForChildWorkflow(
+	nowTimestamp int64,
+) error {
+
+	executionInfo := r.mutableState.GetExecutionInfo()
+	pendingChildWorkflowInfos := r.mutableState.GetPendingChildExecutionInfos()
+
+Loop:
+	for _, childWorkflowInfo := range pendingChildWorkflowInfos {
+		if childWorkflowInfo.StartedID != common.EmptyEventID {
+			continue Loop
+		}
+
+		scheduleEvent, err := r.eventsCache.getEvent(
+			executionInfo.DomainID,
+			executionInfo.WorkflowID,
+			executionInfo.RunID,
+			childWorkflowInfo.InitiatedEventBatchID,
+			childWorkflowInfo.InitiatedID,
+			r.mutableState.GetEventStoreVersion(),
+			r.mutableState.GetCurrentBranch(),
+		)
+		if err != nil {
+			return err
+		}
+
+		if err := r.taskGenerator.generateChildWorkflowTasks(
+			nowTimestamp,
+			scheduleEvent,
+		); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (r *mutableStateTaskRefresherImpl) refreshTasksForCancelExternalWorkflow(
```

Comment:
nit: use RequestCancel insteand of Cancel to be consistent.

Reply:
sure